File: python_leaoai/python_leaoai/services/ApplicationServices/Profile/AccountService.py
import asyncio
from datetime import datetime
from typing import List, Optional
from uuid import uuid4
import hashlib

# --- Models / DTOs ---
class Usuario:
    def __init__(self, id: int = 0, email: str = "", nome: str = "", password: str = "",
                 cpf: str = "", password_token: str = "", password_token_expiry: Optional[datetime] = None):
        self.id = id
        self.email = email
        self.nome = nome
        self.password = password
        self.cpf = cpf
        self.password_token = password_token
        self.password_token_expiry = password_token_expiry

class UserDto:
    def __init__(self, token: str = ""):
        self.token = token

class CreateAccountResponse:
    def __init__(self, email: str, password: str, cpf: str):
        self.email = email
        self.password = password
        self.cpf = cpf

class CreateAccountResult:
    def __init__(self, data: UserDto, message: str):
        self.data = data
        self.message = message

class UpdateAccountResponse:
    def __init__(self, email: str, nome: Optional[str] = None):
        self.email = email
        self.nome = nome

class UpdateAccountResult:
    def __init__(self, success: bool, message: str):
        self.success = success
        self.message = message

class ChangePasswordResponse:
    def __init__(self, current_password: str, new_password: str):
        self.current_password = current_password
        self.new_password = new_password

class ChangePasswordByForgotResponse:
    def __init__(self, guid: str, new_password: str):
        self.guid = guid
        self.new_password = new_password

class UserAlreadyExistsException(Exception): pass
class UserNotFoundException(Exception): pass

# --- Hashing simples ---
class PasswordHasher:
    @staticmethod
    def hash(password: str) -> str:
        return hashlib.sha256(password.encode()).hexdigest()

    @staticmethod
    def verify(hashed: str, password: str) -> bool:
        return hashed == PasswordHasher.hash(password)

# --- Token service simplificado ---
class TokenService:
    @staticmethod
    async def generate_token_async(user: Usuario, service) -> str:
        return str(uuid4())

# --- Repositories fictícios ---
class UsuarioFactory:
    def __init__(self):
        self.users = []

    async def add_usuario_async(self, user: Usuario):
        user.id = len(self.users) + 1
        self.users.append(user)

    def get_user(self, email: str, cpf: str) -> Optional[Usuario]:
        return next((u for u in self.users if u.email == email or u.cpf == cpf), None)

    async def get_user_async(self, email: str) -> Optional[Usuario]:
        return self.get_user(email, "")

    async def update_obj_async(self, user: Usuario):
        for i, u in enumerate(self.users):
            if u.id == user.id:
                self.users[i] = user
                break

    async def get_by_password_reset_guid_async(self, guid: str) -> Optional[Usuario]:
        return next((u for u in self.users if u.password_token == guid), None)

    def get_by_email(self, email: str) -> Optional[Usuario]:
        return next((u for u in self.users if u.email == email), None)

class UserClaim:
    def __init__(self, usuario_id: int, claim_type: str, claim_value: str):
        self.usuario_id = usuario_id
        self.claim_type = claim_type
        self.claim_value = claim_value

class IUserClaimRepository:
    def __init__(self):
        self.claims: List[UserClaim] = []

    async def get_user_claim_async(self, usuario_id: int, claim_type: str):
        return next((c for c in self.claims if c.usuario_id == usuario_id and c.claim_type == claim_type), None)

    async def add_claim_async(self, claim: UserClaim):
        existing = await self.get_user_claim_async(claim.usuario_id, claim.claim_type)
        if existing:
            existing.claim_value = claim.claim_value
        else:
            self.claims.append(claim)

    async def remove_claim_async(self, claim: UserClaim):
        self.claims = [c for c in self.claims if not (c.usuario_id == claim.usuario_id and c.claim_type == claim.claim_type)]

    async def get_user_claims_async(self, usuario_id: int):
        return [c for c in self.claims if c.usuario_id == usuario_id]

    def get_user_claims(self, usuario_id: int):
        return [c for c in self.claims if c.usuario_id == usuario_id]

# --- Serviço de e-mail fictício ---
class IEmailService:
    async def send_password_reset_email_async(self, email, token, *args):
        print(f"Email sent to {email} with token {token}")

# --- Account Service ---
class AccountService:
    def __init__(self, context, repository: UsuarioFactory, user_claim_repository: IUserClaimRepository, email_service: IEmailService):
        self._context = context
        self._repository = repository
        self._user_claim_repository = user_claim_repository
        self._email_service = email_service

    async def create_account(self, model: CreateAccountResponse):
        usuario = self._repository.get_user(model.email, model.cpf)
        if usuario and usuario.id > 0:
            raise UserAlreadyExistsException("Usuário já existe")
        model.password = PasswordHasher.hash(model.password)
        usuario = Usuario(email=model.email, cpf=model.cpf, password=model.password)
        await self._repository.add_usuario_async(usuario)
        token = await TokenService.generate_token_async(usuario, self)
        return CreateAccountResult(data=UserDto(token=token), message="Usuário criado com sucesso")

    async def update_account(self, model: UpdateAccountResponse):
        usuario = await self._repository.get_user_async(model.email)
        if not usuario:
            raise UserNotFoundException("Usuário não encontrado")
        if model.nome:
            usuario.nome = model.nome
        await self._repository.update_obj_async(usuario)
        return UpdateAccountResult(success=True, message="Conta atualizada com sucesso")

    # --- Password ---
    async def forgot_password_async(self, email: str):
        user = await self._repository.get_user_async(email)
        if not user:
            raise Exception("User with the specified email does not exist")
        user.password_token_expiry = datetime.utcnow()
        user.password_token = str(uuid4())
        await self._email_service.send_password_reset_email_async(user.email, user.password_token, "", user.nome, "")

    async def check_token_return_email_async(self, token: str) -> str:
        user = await self._repository.get_by_password_reset_guid_async(token)
        if not user:
            raise Exception("Invalid or expired token")
        return user.email

    async def check_if_password_equals_then_edit_async(self, model: ChangePasswordResponse, user_email: str):
        user = await self._repository.get_user_async(user_email)
        if not user:
            raise Exception("User not found")
        if not PasswordHasher.verify(user.password, model.current_password):
            raise Exception("Current password is incorrect")
        user.password = PasswordHasher.hash(model.new_password)
        await self._repository.update_obj_async(user)

    async def check_if_password_equals_and_guid_async(self, model: ChangePasswordByForgotResponse):
        user = await self._repository.get_by_password_reset_guid_async(model.guid)
        if not user:
            raise Exception("Invalid or expired password reset link")
        if PasswordHasher.verify(user.password, model.new_password):
            raise Exception("The new password must be different from the current password")
        user.password = PasswordHasher.hash(model.new_password)
        await self._repository.update_obj_async(user)

    # --- Claims ---
    async def add_claim_to_user_async(self, usuario_id: int, claim_type: str, claim_value: str):
        existing_claim = await self._user_claim_repository.get_user_claim_async(usuario_id, claim_type)
        if existing_claim:
            existing_claim.claim_value = claim_value
            await self._user_claim_repository.add_claim_async(existing_claim)
        else:
            user_claim = UserClaim(usuario_id, claim_type, claim_value)
            await self._user_claim_repository.add_claim_async(user_claim)

    async def remove_claim_from_user_async(self, usuario_id: int, claim_type: str):
        user_claim = await self._user_claim_repository.get_user_claim_async(usuario_id, claim_type)
        if user_claim:
            await self._user_claim_repository.remove_claim_async(user_claim)

    async def get_user_claims_async(self, usuario_id: int) -> List[dict]:
        claims = await self._user_claim_repository.get_user_claims_async(usuario_id)
        return [{"type": c.claim_type, "value": c.claim_value} for c in claims]

    def get_user_claims(self, usuario_id: int) -> List[dict]:
        claims = self._user_claim_repository.get_user_claims(usuario_id)
        return [{"type": c.claim_type, "value": c.claim_value} for c in claims]

    def get_user_by_email(self, email: str) -> Optional[Usuario]:
        return self._repository.get_by_email(email)

    def encrypt_password(self, password: str):
        return {"hash": PasswordHasher.hash(password)}

File: Framework.Services/ApplicationServices/Profile/AccountService.cs
using Data;
using Data.Models.Pessoa;
using Framework.Data.Models.Users;
using Framework.Domain.Dtos.Common;
using Framework.Domain.Dtos.Profile;
using Framework.Domain.Exceptions;
using Framework.Factories.Pessoa;
using Framework.Infrastructure.Mailing.MessageProvider;
using Framework.Repositories.Interface.Account;
using Framework.Services.Interfaces.Profile;
using Framework.Services.Mapper.Profile;
using SecureIdentity.Password;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;

namespace Framework.Services.ApplicationServices.Profile
{
    public class AccountService : IAccountService
    {
        private readonly ApplicationDbContext _context;
        private readonly UsuarioFactory _repository;
        private readonly IUserClaimRepository _userClaimRepository;
        private readonly IEmailService _emailService;

        public AccountService(ApplicationDbContext context, UsuarioFactory repository, IUserClaimRepository userClaimRepository, IEmailService emailService)
        {
            _context = context;
            _repository = repository;
            _userClaimRepository = userClaimRepository;
            _emailService = emailService;
        }

        public async Task<dynamic> CreateAccount(CreateAccountResponse model)
        {
            Usuario usuario =  _repository.GetUser(model.Email, model.CPF);

            if (usuario is not null && usuario.Id > 0) throw new UserAlreadyExistsException("Usuário já existe");

            model.Password = PasswordHasher.Hash(model.Password);

            usuario = AccountMapper.MappingUser(model);

            await _repository.AddUsuarioAsync(usuario);

            var token = await TokenService.GenerateTokenAsync(usuario, this);

            return new CreateAccountResult (){ Data = new UserDto() { Token = token }, Message = "Usuário criado com sucesso" };
        }

        public async Task<UpdateAccountResult> UpdateAccount(UpdateAccountResponse model)
        {
            // Retrieve the user by email
            var usuario = await _repository.GetUserAsync(model.Email);

            if (usuario == null)
            {
                throw new UserNotFoundException("Usuário não encontrado.");
            }

            // Update user properties only if they are provided
            if (!string.IsNullOrEmpty(model.Nome))
            {
                usuario.Nome = model.Nome;
            }
          
            await _repository.UpdateObjAsync(usuario);

            return new UpdateAccountResult
            {
                Success = true,
                Message = "Conta atualizada com sucesso.",
            };
        }

        public async Task<dynamic> Verification(string email, string token)
        {
            throw new NotImplementedException();
        }

        #region Password
        public async Task ForgotPasswordAsync(string email)
        {
            var user = await _repository.GetUserAsync(email);
            if (user == null)
            {
                throw new Exception("User with the specified email does not exist.");
            }

            user.PasswordTokenExpiry = DateTime.UtcNow;
            user.PasswordToken = Guid.NewGuid().ToString();

            await _emailService.SendPasswordResetEmailAsync(user.Email, user.PasswordToken,"",user.Nome,"");
        }

        public async Task<string> CheckTokenReturnEmailAsync(string token)
        {
            var user = await _repository.GetByPasswordResetGuidAsync(token);
            if (user == null)
            {
                throw new Exception("Invalid or expired token.");
            }
            return user.Email;
        }

        public async Task CheckIfPasswordEqualsThenEditAsync(ChangePasswordResponse model, string userEmail)
        {
            var user = await _repository.GetUserAsync(userEmail);
            if (user == null)
            {
                throw new Exception("User not found.");
            }

            if (!PasswordHasher.Verify(user.Password, model.CurrentPassword))
            {
                throw new Exception("Current password is incorrect.");
            }

            user.Password = PasswordHasher.Hash(model.NewPassword);

            await _repository.UpdateObjAsync(user);
        }

        public async Task CheckIfPasswordEqualsAndGuidAsync(ChangePasswordByForgotResponse model)
        {
            var user = await _repository.GetByPasswordResetGuidAsync(model.Guid);
            if (user == null)
            {
                throw new Exception("Invalid or expired password reset link.");
            }

            if (PasswordHasher.Verify(user.Password, model.NewPassword))
            {
                throw new Exception("The new password must be different from the current password.");
            }

            user.Password = PasswordHasher.Hash(model.NewPassword);

            await _repository.UpdateObjAsync(user);
        }
        #endregion

        #region Claims
        public async Task AddClaimToUserAsync(int usuarioId, string claimType, string claimValue)
        {
            var existingClaim = await _userClaimRepository.GetUserClaimAsync(usuarioId, claimType);
            if (existingClaim != null)
            {
                // Update the claim value
                existingClaim.ClaimValue = claimValue;
                await _userClaimRepository.AddClaimAsync(existingClaim);
            }
            else
            {
                // Add a new claim
                var userClaim = new UserClaim
                {
                    UsuarioId = usuarioId,
                    ClaimType = claimType,
                    ClaimValue = claimValue
                };
                await _userClaimRepository.AddClaimAsync(userClaim);
            }
        }

        public async Task RemoveClaimFromUserAsync(int usuarioId, string claimType)
        {
            var userClaim = await _userClaimRepository.GetUserClaimAsync(usuarioId, claimType);
            if (userClaim != null)
            {
                await _userClaimRepository.RemoveClaimAsync(userClaim);
            }
        }

        public async Task<List<Claim>> GetUserClaimsAsync(int usuarioId)
        {
            var userClaims = await _userClaimRepository.GetUserClaimsAsync(usuarioId);
            return userClaims.Select(uc => new Claim(uc.ClaimType, uc.ClaimValue)).ToList();
        }

        public List<Claim> GetUserClaims(int usuarioId)
        {
            var userClaims = _userClaimRepository.GetUserClaims(usuarioId);
            return userClaims.Select(uc => new Claim(uc.ClaimType, uc.ClaimValue)).ToList();
        }

        public Usuario GetUserByEmail(string emailAddress)
        {
            return _repository.GetByEmail(emailAddress);
        }

        public object EncryptPassword(string password)
        {
            return new { hash = PasswordHasher.Hash(password) };
        }

        #endregion
    }
}
